File: api/src/main/java/com/kagg886/sylu_eoa/model/ClassUnit.java
package com.kagg886.sylu_eoa.model;

import com.alibaba.fastjson2.annotation.JSONField;
import lombok.Data;
import lombok.ToString;

import java.util.ArrayList;
import java.util.List;

/**
 * 无
 *
 * @author kagg886
 * @date 2023/9/4 19:55
 **/
@Data
@ToString
public class ClassUnit {
    public static ClassUnit EMPTY = new ClassUnit(null, null, null, null, null, "0");

    private String name;
    private String teacher;
    private String room;
    private String lesson;
    private int dayInWeek;
    private List<Range> weekAsMinMax;
    private String weekEachLesson;

    public ClassUnit(
            @JSONField(name = "name") String name,
            @JSONField(name = "teacher") String teacher,
            @JSONField(name = "room") String room,
            @JSONField(name = "weekEachLesson") String weekEachLesson,
            @JSONField(name = "lesson") String lesson,
            @JSONField(name = "dayInWeek") String dayInWeek
//            @JSONField(name = "kcmc") String name,
//            @JSONField(name = "xm") String teacher,
//            @JSONField(name = "cdmc") String room,
//            @JSONField(name = "zcd") String weekEachLesson,
//            @JSONField(name = "jcs") String lesson,
//            @JSONField(name = "xqj") String dayInWeek
    ) {
        if (name == null) {
            return;
        }
        this.name = name;
        this.teacher = teacher;
        this.room = room;
        this.lesson = lesson;
        this.dayInWeek = Integer.parseInt(dayInWeek);
        this.weekEachLesson = weekEachLesson;

        List<Range> rtn = new ArrayList<>();
        for (String a : weekEachLesson.split(",")) {
            a = a.substring(0, a.length() - 1);
            if (a.contains("-")) {
                String[] k = a.split("-");
                int l;
                FilterType type = FilterType.ALL;
                try {
                    l = Integer.parseInt(k[1]);
                } catch (NumberFormatException e) {
                    l = Integer.parseInt(k[1].split("周")[0]);
                    switch (k[1].split("\\(")[1]) {
                        case "单" -> type = FilterType.SINGULAR;
                        case "双" -> type = FilterType.EVEN;
                    }
                }
                rtn.add(new Range(Integer.parseInt(k[0]), l, type));
                continue;
            }
            rtn.add(new Range(Integer.parseInt(a), Integer.parseInt(a), FilterType.ALL));
        }
        this.weekAsMinMax = rtn;
    }


    @Data
    @ToString
    public static class Range {
        private int start;
        private int end;
        private FilterType type;

        public Range(int start, int end, FilterType type) {
            this.start = start;
            this.end = end;
            this.type = type;
        }

        public Range() {

        }

        public static String formatToString(Range r) {
            StringBuilder builder = new StringBuilder();
            builder.append("第").append(r.start);
            if (r.start == r.end) {
                return builder.append("周").toString();
            }
            builder.append("周---第").append(r.end).append("周");
            if (r.getType() != FilterType.ALL) {
                builder.append("(").append(r.getType() == FilterType.SINGULAR ? "单" : "双").append("周)");
            }
            return builder.toString();
        }
    }

    public enum FilterType {
        ALL, //单双周
        SINGULAR,//单周
        EVEN;//双周
    }
}

File: eoa-lib/network-core/src/commonMain/kotlin/top/kagg886/sylu_eoa/api/v2/bean/ClassUnit.kt
package top.kagg886.sylu_eoa.api.v2.bean

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class ClassUnit(
    //名字
    @SerialName("kcmc") val name: String,
    //老师名字
    @SerialName("xm") val teacher: String,
    //房间
    @SerialName("cdmc") val room: String,
    //第几周有课
    @SerialName("zcd") val weekEachLesson: String,
    //节数
    @SerialName("jcs") val lesson: String,
    //星期几 1 2 3 4 5 6 7
    @SerialName("xqj") val dayInWeek: String,

    //学分
    @SerialName("xf") val score: String,

    //考察形式（考查，考试）
    @SerialName("khfsmc") val classType: String,

    @SerialName("zyhxkcbj") private val _degreeProgram: String,
) {
    val isDegreeProgram by lazy {
        _degreeProgram == "是"
    }
    //1-2节
    val rangeEveryDay by lazy {
        val ls = lesson.split("-")
        ((ls[0]).toInt() ..(ls[1]).toInt()).toList()
    }

    //7周,9-11周(单),12-16周, 一定大于1。
    val rangeAllTerm by lazy {
        weekEachLesson.replace("周", "").split(",").map {
            val a = it.substring(0, it.length)
            if (!a.contains("-")) {
                return@map listOf(a.toInt())
            }
            val range = a.split("-")

            var end = range[1]
            var step = 1
            if (end.contains("(")) {
                step = 2
                end = end.substring(0, end.indexOf("("))
            }
            (range[0].toInt()..end.toInt() step step).toList()
        }.flatten()
    }
}

fun List<ClassUnit>.findClassByWeek(
    weekForSemester: Int,
): List<ClassUnit> {
    return filter {
        it.rangeAllTerm.contains(weekForSemester)
    }
}